x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java

```diff
@@ -58,28 +65,39 @@ protected void doExecute(
         InferTrainedModelDeploymentAction.Request request,
         ActionListener<InferTrainedModelDeploymentAction.Response> listener
     ) {
-        String deploymentId = request.getDeploymentId();
-        // We need to check whether there is at least an assigned task here, otherwise we cannot redirect to the
-        // node running the job task.
-        TrainedModelAllocation allocation = TrainedModelAllocationMetadata.allocationForModelId(
-            clusterService.state(),
-            request.getDeploymentId()
-        ).orElse(null);
-        if (allocation == null) {
-            String message = "Cannot perform requested action because deployment [" + deploymentId + "] is not started";
-            listener.onFailure(ExceptionsHelper.conflictStatusException(message));
-            return;
-        }
-        String[] randomRunningNode = allocation.getStartedNodes();
-        if (randomRunningNode.length == 0) {
-            String message = "Cannot perform requested action because deployment [" + deploymentId + "] is not yet running on any node";
-            listener.onFailure(ExceptionsHelper.conflictStatusException(message));
-            return;
-        }
-        // TODO Do better routing for inference calls
-        int nodeIndex = Randomness.get().nextInt(randomRunningNode.length);
-        request.setNodes(randomRunningNode[nodeIndex]);
-        super.doExecute(task, request, listener);
+        final String deploymentId = request.getDeploymentId();
+        provider.getTrainedModel(deploymentId, GetTrainedModelsAction.Includes.empty(), ActionListener.wrap(config -> {
```

Comment:
This is quite expensive to be doing on each inference request. Could we make this work by just reading the allocation metadata?

Reply:
> Could we make this work by just reading the allocation metadata?

I agree, if `_infer` is to be used in a hot-path, we shouldn't grab the model everytime. The trick is, using the deployment metadata only tells us if the deployment exists or not, not if the model exists or not. 

I will think on this again, we could simply update the error to be a nicer 404 message indicating the deployment and/or the model doesn't exist.

Reply:
@dimitris-athanasiou how about this, if the deployment is missing, we THEN check if the model exists or not to provide a nicer message?

This way, the `_infer` path is fast when the deployment exists.

Reply:
Yeah, I was thinking we could do this too. We could be potentially making ingest slower but only scenarios where there is a clear misconfiguration for the user's pipelines.